js/src/components/shipping-rate-section/minimum-order-card/add-minimum-order-modal.js

```diff
@@ -0,0 +1,110 @@
+/**
+ * External dependencies
+ */
+import { Button } from '@wordpress/components';
+import { __ } from '@wordpress/i18n';
+import { Form } from '@woocommerce/components';
+import { noop } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import AppModal from '.~/components/app-modal';
+import AppInputPriceControl from '.~/components/app-input-price-control/index.js';
+import VerticalGapLayout from '.~/components/vertical-gap-layout';
+import AppCountrySelect from '.~/components/app-country-select';
+
+const AddMinimumOrderModal = ( props ) => {
+	const { countryOptions, value, onChange = noop, onRequestClose } = props;
+
+	const handleValidate = ( values ) => {
+		const errors = {};
+
+		if ( values.countries.length === 0 ) {
+			errors.countries = __(
+				'Please specify at least one country.',
+				'google-listings-and-ads'
+			);
+		}
+
+		if ( values.threshold < 0 ) {
+			errors.price = __(
+				'The minimum order amount cannot be less than 0.',
+				'google-listings-and-ads'
+			);
+		}
+
+		return errors;
+	};
+
+	const handleSubmitCallback = ( newValue ) => {
+		onRequestClose();
+		onChange( newValue );
+	};
+
+	return (
+		<Form
+			initialValues={ {
+				countries: value.countries,
+				threshold: value.threshold,
+				currency: value.currency,
```

Comment:
I want the currency to go side-by-side together with the threshold value. I think that is a better design in the long term, in case there are scenarios like different thresholds with different currencies, this would be able to work with that. Also, previously there was a related issue https://github.com/woocommerce/google-listings-and-ads/issues/1189 in shipping rate, where we were only looking at number without currency. I'd like to avoid that kind of situation.

Currency is actually used as `values.currency` in `AppInputPriceControl`'s `suffix` props.

Reply:
❔ 

> I think that is a better design in the long term, in case there are scenarios like different thresholds with different currencies, this would be able to work with that.

If the currency is different, there is still no need to put it in `initialValues`. If consider that the currency could be edited in the future, then it would be more obvious at that time whether to put it in `initialValues` or to use another way to handle it when the actual need appears.

When reading codes, it is a bit confusing to see that `currency` is not a form item and does not need to be passed down to any separated sub-components, but only to be passed to the `values` below and then placed in the `suffix` prop.

> Also, previously there was a related issue #1189 in shipping rate, where we were only looking at number without currency. I'd like to avoid that kind of situation.

I couldn't understand why dose placing `currency` in `initialValues` relate to issue #1189? 🤔 

> Currency is actually used as `values.currency` in `AppInputPriceControl`'s `suffix` props.

It could be changed to `suffix={ value.currency }`.

Reply:
> When reading codes, it is a bit confusing to see that `currency` is not a form item and does not need to be passed down to any separated sub-components...

I guess your understanding or expectation is that when we see the Form `initialValues`, the values should have a corresponding UI that can be seen and/or modified by the users, and in code level, there should be a corresponding `{ ...getInputProps( 'fieldname' ) }` for the value?

My understanding on Form `initialValues` is more than that though. I see it as a logical grouping of things. In some cases, it can be used as a Form "state" without needing to use `useState`. The values may or may not have a corresponding UI or a corresponding `{ ...getInputProps( 'fieldname' ) }`. Sometimes a value may be a derived value based on other values (e.g. using `setValue`), and we want to group them together. Sometimes a value may participate in validation.

I gave some thoughts about your comment: `currency` is not modifiable by users, there is no `setValue` for it, and it does not participate in validation. I proceeded to remove `currency` from `initialValues`, and when I was trying it out, I realized that because `currency` is no longer grouped together, I had to make a bit modification to `handleSubmitCallback` to achieve the same outcome:

```js
// (Existing) Approach 1: newValue has countries, threshold and currency.
const handleSubmitCallback = ( newValue ) => {
	onRequestClose();
	onChange( newValue );
};

// (Suggested) Approach 2. newValue has countries and threshold, no currency.
const handleSubmitCallback = ( newValue ) => {
	onRequestClose();
	onChange( {
		... newValue,
		currency: value.currency
	} );
};
```

When I look at the above, I prefer Approach 1. It relates to my point on logical grouping of things. The shape of `value` is consistent throughout and requires less mental process, in my opinion.

Keen to hear your opinion on this. 😄  

> ... but only to be passed to the `values` below and then placed in the `suffix` prop.

> It could be changed to `suffix={ value.currency }`.

Because `currency` _is already in the form values_, the child should then get the value from the Form `values` instead of the outer `value` in the AddMinimumOrderModal component. This would make things easier in case we want to do refactoring on the child code.

> > Also, previously there was a related issue #1189 in shipping rate, where we were only looking at number without currency. I'd like to avoid that kind of situation.
> 
> I couldn't understand why dose placing `currency` in `initialValues` relate to issue #1189? 🤔

They are _not technically code related_. What I meant to say is the _idea_ to look at _the price (the number and the currency) as a logical group_. (Sorry 😄 🙏 )

Reply:
Thanks for the explanation!

> I guess your understanding or expectation is that when we see the Form `initialValues`, the values should have a corresponding UI that can be seen and/or modified by the users, and in code level, there should be a corresponding `{ ...getInputProps( 'fieldname' ) }` for the value?

> My understanding on Form `initialValues` is more than that though. I see it as a logical grouping of things.

From my understanding, the `initialValues` is more likely the values that are used closely with form and its fields. Even if there are other logically related (grouping) values, adding them might still be relatively not ideal if there are no special situations.

Because the event and state handling of forms are often relatively complex, using it as a logical grouping of things might increase the complexity.

> In some cases, it can be used as a Form "state" without needing to use `useState`.

Sorry I'm not quite sure whether borrowing a \<Form\> value as a state is a good idea. It will trigger \<Form\>'s `onChange` callback when calling `setValue`, which makes callback to be triggered when there is no "real" form values changed.

And that state will also appear in the arguments of `validate` and `onSubmit` callbacks. The mix of form values and states in \<Form\> callbacks could be confusing when reading the code, further reducing readability and complexity when maintaining it.

> The values may or may not have a corresponding UI or a corresponding `{ ...getInputProps( 'fieldname' ) }`.

It's useful when values have corresponding `{ ...getInputProps( 'fieldname' ) }`. But sorry I'm not so sure why it's useful when the values may or may not have a corresponding UI? Seems they are more like props.

> Sometimes a value may be a derived value based on other values.

If there is a derived value based on form values, It could be handled in \<Form\> `onChange`. In this way, it provides a much clear overview to figure out how this derived value is updated accordingly, and lower coupling.

> Sometimes a value may participate in validation.

This is indeed more convenient, but I believe if a logical grouping value that participates in validation could be put in `initialValues`, then usually it was already accessible in `validate` callback as well. So, when their relationship is in validation scope, would it be a bit better to make them logically grouped in the `validate` callback rather than the `initialValues`? 

---

> I proceeded to remove `currency` from `initialValues`, and when I was trying it out, I realized that because `currency` is no longer grouped together, I had to make a bit modification to `handleSubmitCallback` to achieve the same outcome:
> ```js
> // ... omitted
> // (Suggested) Approach 2. newValue has countries and threshold, no currency.
> const handleSubmitCallback = ( newValue ) => {
> 	onRequestClose();
> 	onChange( {
> 		... newValue,
> 		currency: value.currency
> 	} );
> };
>
> ```
> When I look at the above, I prefer Approach 1. It relates to my point on logical grouping of things. The shape of value is consistent throughout and requires less mental process, in my opinion.
> 
> Keen to hear your opinion on this.

Sorry, I'm not quite sure why does the same shape of `value` here can requires less mental process nor how to evaluate when things have logical grouping relation. In my experience, adding JSDocs for them might be more intuitive and effective in reducing mental processes.

And I think if removing `currency` was ever a choice to be used, then maintaining a consistent shape should no longer be a need since they are already inconsistent in the `initialValues`. And because `currency` is not on any of above examples listed for logical grouping, then maintaining logical grouping should not be a consideration either.

I commented whether `currency` could be removed at a few places in this PR because I went back and forth several times between these files during the first code review, to see where `currency` was being used. To think about whether there are any edge cases where `currency` is needed that I didn't notice. And this process was repeated twice when I replied to this comment.

After repeatedly verifying that `currency` just needs to be passed to `suffix`, I thought deleting them could improve readability and maintainability. Otherwise, future maintenance might still cause similar back-and-forth and validation processes as above.

> Because `currency` _is already in the form values_, the child should then get the value from the Form `values` instead of the outer `value` in the AddMinimumOrderModal component. This would make things easier in case we want to do refactoring on the child code.

Sorry, I'm not sure I agree that this is a "should" convention/principle because it seems a bit of reverse causation, and the `value.currency` could even be a `suffix` prop of \<AddMinimumOrderModal\> rather than mixing in `value` prop. Based on above comments, binding it with `value` prop makes it be confusing, in my opinion.

And if how easier when refactoring is a concern at this point, then it could be handled in the same way as `countryOptions` and `onRequestClose` when it happens.

---

After the above discussion, I knew that there are some reasons for keeping the same shape/logical grouping. Although I have some concerns about readability and maintainability, that is just my subjective opinion.

Given that the implementation is workable, and that my comment started with a 💅 . I think this PR can be moved on regardless of which option is chosen.

Reply:
Thanks so much for the explanation! 

> I commented whether `currency` could be removed at a few places in this PR because I went back and forth several times between these files during the first code review, to see where `currency` was being used. To think about whether there are any edge cases where `currency` is needed that I didn't notice. And this process was repeated twice when I replied to this comment.

I think this one above is especially noteworthy for me. Now I know your perspective better. I got to let it sink in my mind for a bit and see how it turns out. 😄